src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs

```diff
@@ -0,0 +1,616 @@
+    private readonly OutputCacheOptions _options;
+    private readonly ILogger _logger;
+    private readonly IOutputCacheStore _store;
+    private readonly IOutputCacheKeyProvider _keyProvider;
+    private readonly WorkDispatcher<string, OutputCacheEntry?> _outputCacheEntryDispatcher;
+    private readonly WorkDispatcher<string, OutputCacheEntry?> _requestDispatcher;
+
+    /// <summary>
+    /// Creates a new <see cref="OutputCacheMiddleware"/>.
+    /// </summary>
+    /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+    /// <param name="options">The options for this middleware.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> used for logging.</param>
+    /// <param name="outputCache">The <see cref="IOutputCacheStore"/> store.</param>
+    /// <param name="poolProvider">The <see cref="ObjectPoolProvider"/> used for creating <see cref="ObjectPool"/> instances.</param>
+    public OutputCacheMiddleware(
+        RequestDelegate next,
+        IOptions<OutputCacheOptions> options,
+        ILoggerFactory loggerFactory,
+        IOutputCacheStore outputCache,
+        ObjectPoolProvider poolProvider
+        )
+        : this(
+            next,
+            options,
+            loggerFactory,
+            outputCache,
+            new OutputCacheKeyProvider(poolProvider, options))
+    { }
+
+    // for testing
+    internal OutputCacheMiddleware(
+        RequestDelegate next,
+        IOptions<OutputCacheOptions> options,
+        ILoggerFactory loggerFactory,
+        IOutputCacheStore cache,
+        IOutputCacheKeyProvider keyProvider)
+    {
+        ArgumentNullException.ThrowIfNull(next);
+        ArgumentNullException.ThrowIfNull(options);
+        ArgumentNullException.ThrowIfNull(loggerFactory);
+        ArgumentNullException.ThrowIfNull(cache);
+        ArgumentNullException.ThrowIfNull(keyProvider);
+
+        _next = next;
+        _options = options.Value;
+        _logger = loggerFactory.CreateLogger<OutputCacheMiddleware>();
+        _store = cache;
+        _keyProvider = keyProvider;
+        _outputCacheEntryDispatcher = new();
+        _requestDispatcher = new();
+    }
+
+    /// <summary>
+    /// Invokes the logic of the middleware.
+    /// </summary>
+    /// <param name="httpContext">The <see cref="HttpContext"/>.</param>
+    /// <returns>A <see cref="Task"/> that completes when the middleware has completed processing.</returns>
+    public async Task Invoke(HttpContext httpContext)
+    {
+        // Skip the middleware if there is no policy for the current request
+        if (!TryGetRequestPolicies(httpContext, out var policies))
+        {
+            await _next(httpContext);
+            return;
+        }
+
+        var context = new OutputCacheContext(httpContext, _store, _options, _logger);
+
+        // Add IOutputCacheFeature
+        AddOutputCacheFeature(context);
```

Comment:
Is the feature needed when `context.EnableOutputCaching` is `false` below?

Reply:
Calling `await policy.CacheRequestAsync(context);` will alter the context and the value of this property.